### DISCUSSION TITLE
multivariate time series classification - how-to

### DISCUSSION THREAD

**Discussion by irazza** [alan-turing-institute/sktime#4786] [Q&A] [answered]
I'm currently working on a multi-modal binary classification problem and would greatly appreciate some guidance and assistance. I have two input signals for every observation and need to predict a single binary label as the output.

I've already explored the documentation and examples provided by `sktime`, but I'm unsure about the specific steps and techniques to use for this particular scenario. I'm reaching out to the community to seek advice and suggestions on how to approach this problem effectively using `sktime`.

Here are some specific questions I have:

1. How can I represent the multi-modal input signals in the required format for `sktime`? Should I reshape my data manually or are there specific functions available for this purpose?

2. Which classifier in `sktime` would be suitable for this multi-modal binary classification problem? Are there any specific considerations or parameters that I should be aware of?

3. What are the best practices or recommended workflows for preprocessing and feature extraction in a multi-modal context using `sktime`?

I would greatly appreciate any insights, code snippets, or references to relevant resources that can help me tackle this problem effectively. If you have previously worked on similar multi-modal classification tasks with `sktime`, I would love to hear about your experiences and any lessons learned.

Thank you in advance for your time and assistance. I'm looking forward to your valuable input and guidance!

**Comment by fkiraly** [comment, marked as answer]
Welcome, @albertoazzari!

I think what you are referring to is the same as multivariate binary classification? That is, you have n trials (e.g., athletes?), and in each trial you measure m modalities (speed, position, etc) over a number of time points that may depend on the trial?

If so, then our recent tutorial at pydata London 2023 should have all the information you need:
https://github.com/sktime/sktime-tutorial-pydata-london-2023

Notebook 2 give an introduction to the general data format for multivariate classification, and the learning task itself.

The notebook on data loading, in section 4, also contains some common patterns for loading data from csv:
https://github.com/sktime/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb

This should help with question 1.

**Reply by fkiraly** [reply]
Re 2, you can search `sktime` using `all_estimators` from the `registry` module:

```python
from sktime.registry import all_estimators

all_estimators("classifier", filter_tags={"capability:multivariate": True}, as_dataframe=True)
```

lists all estimators or composers that can deal with multivariate data.

**Reply by fkiraly** [reply]
Re 3, best practice recommendations differ by the person who makes them, so you may like to consult academic literature.

Some general best practice:

* always have a baseline to compare - in this case that would be `DummyClassifier`
* in classification, don't mix pieces of instances in train/test set, keep entire instances/trials out
* if you do manual model development, keep a "final" test set (e.g., 20 or 30%) out until the very end when you have settled on a model. Developing complex models on the same batch or collection of data is a common mistake that is hard to spot, it is sometimes called "graduate student descent".
* settle on metrics for performance evaluation *before* you run a benchmarking experiment
* when comparing performances, ensure to get your statistical testing right. Common points of confusion is between measures in single data set and multiple data sets, and performance of an algorithm as fitted vs performance of an algorithm including fitting.

Regarding best practice in method choice, opinions diverge - here's common algorithms you can try:

* naive baseline should always be included :-)
* estimators of the kind "simple feature extraction (e.g., `SummaryTransformer`) then apply a tabular classifier (e.g., `RandomForest`, `SVC`)", if you end up with a specialist time series classifier, it should outperform these
    * specialist feature extractors, e.g., tsfresh, rocket, matrixprofile are popular - these can be combined with `sklearn` classifiers (tabular) or `sktime` classifiers
* basic classifier using distances: k-nearest neighbor with time series distance (`KNeighborsTimeSeriesClassifier`). The common choice is combining knn with dynamic time warping or other common time series distances. Simple baseline choices are average/flattened distances (e.g., Euclidan distance on the time series as vetor).
* basic classifier using kernels: support vector classifier (`TimeSeriesSVC`). The common choice is an alignment kernel. Simple baseline choices are average/flattened kernel (e.g., Gaussian kernel on the time series as vector).
    * for multivariate classification with distances/kernels, you can either use a multivariate distance (e.g., dtw with multivariate inner distance), or use a univariate distance and then use one of the techniques below to get a multivariate classifier
    * you can list time series distances and kernels with `all_estimators("transformer-pairwise-panel", etc)`
* the above (especially svc, knn), often can be improved by tuning via grid search, possibly accompanied by feature extraction that is specific to your data.
* there are general techniques on turning any univariate classifier into a multivariate one, the most common ones are column ensembling via `ColumnEnsembleClassifier`. Bagging also works, via `BaggingClassifier`, although I just noticed this has a bug (fixed but not until 0.20.1)
    * if the classifier is distance or kernel-based, you can get a multivariate distance/kernel from a univariate one using sum-of- or product-of variables. This can be done, e.g., via the `IndepDist` compositior.
* Beyond this, there are "one size fits all" multivariate classifiers, and every academic network has their favourites. You will likely get different recommendations here depending on whom you ask - in the table these are usually the classifiers with catchy proper names, e.g., `DrCIF`, `FreshPRINCE`, `HIVECOTEV2`, and so on. This can be hit & miss...

The list above is roughly in order of how I personally would try. As said, this may differ from recommendations by others.
